File: MATLAB/src/utility/meshSuperquadrics.m
function [mesh_merge] = meshSuperquadrics(x, varargin)

n = size(x, 1);

% with tapering or not
taper = false;
% color = 'r';
arclength = 0.02;

for k = 1 : size(varargin, 2)
    if strcmp(varargin{k}, 'Taper')
        taper = varargin{k + 1};
    end
%     if strcmp(varargin{k}, 'Color')
%         color = varargin{k + 1};
%     end   
    if strcmp(varargin{k}, 'Arclength')
        arclength= varargin{k + 1};
    end    
end

mesh = cell(n, 1);
v_num = zeros(n, 1);
f_num = zeros(n, 1);


for i = 1 : n
    [x_mesh, y_mesh, z_mesh] = singleMeshSuperquadrics(x(i, :), ...
        'Arclength', arclength, 'Taper', taper);
    [mesh{i}.f, mesh{i}.v] = mesh2tri(x_mesh, y_mesh, z_mesh,'b'); % f b x
    v_num(i) = size(mesh{i}.v, 1);
    f_num(i) = size(mesh{i}.f, 1);
    if i > 1
        mesh{i}.f = mesh{i}.f + sum(v_num(1 : i - 1));
    end
end
mesh_merge.v = zeros(sum(v_num), 3);
mesh_merge.f = zeros(sum(f_num), 3);
mesh_merge.v(1 : v_num(1), :) = mesh{1}.v;
mesh_merge.f(1 : f_num(1), :) = mesh{1}.f;
for i = 2 : n
    mesh_merge.v(...
        sum(v_num(1 : i - 1)) + 1 : sum(v_num(1 : i - 1)) + v_num(i), :) = ...
        mesh{i}.v;
    mesh_merge.f(...
        sum(f_num(1 : i - 1)) + 1 : sum(f_num(1 : i - 1)) + f_num(i), :) = ...
        mesh{i}.f;
end
end

File: src/marching_primitives/mesh_superquadrics.py
import numpy as np
from .single_mesh_superquadrics import single_mesh_superquadrics
from .mesh2tri import mesh2tri


def mesh_superquadrics(x, arclength=0.02, taper=False):
    """
    Convert multiple superquadric parameters to a merged triangulated mesh.

    Parameters
    ----------
    x : ndarray (K, 11) or (K, 13)
        Superquadric parameters, one per row.
    arclength : float
        Arclength sampling resolution.
    taper : bool
        Whether tapering is enabled.

    Returns
    -------
    mesh : dict
        Dictionary with 'v' (vertices, Nx3) and 'f' (faces, Mx3) keys.
    """
    n = x.shape[0]
    meshes = []

    v_offset = 0
    all_v = []
    all_f = []

    for i in range(n):
        x_mesh, y_mesh, z_mesh = single_mesh_superquadrics(
            x[i], arclength=arclength, taper=taper
        )
        f, v = mesh2tri(x_mesh, y_mesh, z_mesh, 'f')

        # Remove mesh sections (matching MATLAB: idx = s/8 : 3s/8, 5s/8 : 7s/8)
        s = f.shape[0]
        idx1 = np.arange(s // 8, 3 * s // 8)
        idx2 = np.arange(5 * s // 8, 7 * s // 8)
        idx = np.concatenate([idx1, idx2])
        f = f[idx]

        # Offset face indices
        f = f + v_offset
        v_offset += v.shape[0]

        all_v.append(v)
        all_f.append(f)

    mesh = {
        'v': np.vstack(all_v),
        'f': np.vstack(all_f)
    }
    return mesh
